scheduler/service/service.go

```diff
@@ -379,112 +380,116 @@ func (s *Service) LeaveTask(ctx context.Context, req *rpcscheduler.PeerTarget) e
-	if !loaded || task == nil {
-		msg := "task not found in P2P network"
-		log.Info(msg)
+	if !loaded {
+		msg := fmt.Sprintf("task %s not found", req.TaskId)
+		logger.Info(msg)
+		// TODO(eryugey) use dferrors.Newf
 		return common.NewGrpcDfResult(base.Code_PeerTaskNotFound, msg), nil
 	}
+
 	if task.FSM.Current() != resource.TaskStateSucceeded {
-		msg := fmt.Sprintf("task found but not in %s state: %s", resource.TaskStateSucceeded, task.FSM.Current())
-		log.Info(msg)
+		msg := fmt.Sprintf("task has been found but state is %s", task.FSM.Current())
+		task.Log.Info(msg)
+		// TODO(eryugey) use dferrors.Newf
 		return common.NewGrpcDfResult(base.Code_PeerTaskNotFound, msg), nil
 	}
 
-	lenPeers := task.PeerCount.Load()
-	if lenPeers <= 0 {
-		msg := fmt.Sprintf("task found but with %d active peers", lenPeers)
-		log.Info(msg)
+	if !task.HasAvailablePeer() {
+		msg := "task has been found but task has no available peers"
+		task.Log.Info(msg)
+		// TODO(eryugey) use dferrors.Newf
 		return common.NewGrpcDfResult(base.Code_PeerTaskNotFound, msg), nil
 	}
 
-	msg := "task found in P2P network"
-	log.Info(msg)
-	return common.NewGrpcDfResult(base.Code_Success, msg), nil
+	// TODO(eryugey) use dferrors.Newf
+	return common.NewGrpcDfResult(base.Code_Success, "task has been found"), nil
 }
 
 // AnnounceTask informs scheduler a peer has completed task
 func (s *Service) AnnounceTask(ctx context.Context, req *rpcscheduler.AnnounceTaskRequest) (*base.GrpcDfResult, error) {
-	log := logger.With("function", "AnnounceTask", "TaskID", req.TaskId, "CID", req.Cid)
-
-	// Must contain valid PiecePacket.PieceInfos
-	if req.PiecePacket == nil || req.PiecePacket.PieceInfos == nil {
-		msg := "no piece info found in request"
-		log.Warn(msg)
-		return common.NewGrpcDfResult(base.Code_BadRequest, msg), nil
-	}
-
+	taskID := req.TaskId
+	peerID := req.PiecePacket.DstPid
 	pieceInfos := req.PiecePacket.PieceInfos
 	totalPiece := req.PiecePacket.TotalPiece
-	contentLength := req.PiecePacket.ContentLength
-	peerID := req.PiecePacket.DstPid
-	if totalPiece != int32(len(pieceInfos)) {
-		msg := fmt.Sprintf("total piece count mismatch in request: %d != %d", totalPiece, len(pieceInfos))
-		log.Warn(msg)
-		return common.NewGrpcDfResult(base.Code_BadRequest, msg), nil
-	}
-
-	// Check if taskID is valid
-	if req.TaskId != req.PiecePacket.TaskId {
-		msg := fmt.Sprintf("task ID mismatch %s != %s", req.TaskId, req.PiecePacket.TaskId)
-		log.Warn(msg)
+	if len(pieceInfos) <= 0 || totalPiece != int32(len(pieceInfos)) {
+		msg := fmt.Sprintf("length of pieces is invalid, total piece id %d, length of piece infos is %d", totalPiece, len(pieceInfos))
+		logger.WithTaskAndPeerID(taskID, peerID).Warn(msg)
+		// TODO(eryugey) use dferrors.Newf
 		return common.NewGrpcDfResult(base.Code_BadRequest, msg), nil
 	}
 
-	task := resource.NewTask(req.TaskId, req.Cid, 0, req.UrlMeta)
+	task := resource.NewTask(taskID, req.Cid, s.config.Scheduler.BackSourceCount, req.UrlMeta)
```

Comment:
Cache can't do backsource, so I set backsource to 0 on purpose, as in cache mode, app does backsource not dfdaemon nor cdn.